amethyst_network/src/simulation/transport/tcp.rs

```diff
@@ -0,0 +1,276 @@
+impl<'a, 'b> SystemBundle<'a, 'b> for TcpNetworkBundle {
+    fn build(
+        self,
+        world: &mut World,
+        builder: &mut DispatcherBuilder<'_, '_>,
+    ) -> Result<(), Error> {
+        builder.add(
+            NetworkSimulationTimeSystem,
+            NETWORK_SIM_TIME_SYSTEM_NAME,
+            &[],
+        );
+        builder.add(
+            TcpConnectionListenerSystem,
+            CONNECTION_LISTENER_SYSTEM_NAME,
+            &[NETWORK_SIM_TIME_SYSTEM_NAME],
+        );
+        builder.add(
+            TcpNetworkSendSystem,
+            NETWORK_SEND_SYSTEM_NAME,
+            &[CONNECTION_LISTENER_SYSTEM_NAME],
+        );
+        builder.add(
+            TcpNetworkRecvSystem::new(),
+            NETWORK_RECV_SYSTEM_NAME,
+            &[CONNECTION_LISTENER_SYSTEM_NAME],
+        );
+        world.insert(TcpNetworkResource::new(
+            self.listener,
+            self.recv_buffer_size_bytes,
+        ));
+        Ok(())
+    }
+}
+
+/// System to listen for incoming connections and cache them to the resource.
+pub struct TcpConnectionListenerSystem;
+
+impl<'s> System<'s> for TcpConnectionListenerSystem {
+    type SystemData = (
+        Write<'s, TcpNetworkResource>,
+        Read<'s, NetworkSimulationTime>,
+        Write<'s, EventChannel<NetworkSimulationEvent>>,
+    );
+
+    fn run(&mut self, (mut net, sim_time, mut event_channel): Self::SystemData) {
+        let resource = net.deref_mut();
+        if let Some(ref listener) = resource.listener {
+            loop {
+                match listener.accept() {
+                    Ok((stream, addr)) => {
+                        stream
+                            .set_nonblocking(true)
+                            .expect("Setting nonblocking mode");
+                        stream.set_nodelay(true).expect("Setting nodelay");
+                        resource.streams.insert(addr, stream);
+                        event_channel.single_write(NetworkSimulationEvent::Connect(addr));
+                    }
+                    Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                        break;
+                    }
+                    Err(e) => {
+                        error!("Error listening for connections: {}", e);
+                        break;
+                    }
+                };
+            }
+        }
+    }
+}
+
+/// System to send messages to a particular open `TcpStream`.
+pub struct TcpNetworkSendSystem;
+
+impl<'s> System<'s> for TcpNetworkSendSystem {
+    type SystemData = (
+        Write<'s, TransportResource>,
+        Write<'s, TcpNetworkResource>,
+        Read<'s, NetworkSimulationTime>,
+    );
+
+    fn run(&mut self, (mut transport, mut net, sim_time): Self::SystemData) {
+        let messages = transport.messages_to_send(|_| sim_time.should_send_messages());
+        for message in messages.iter() {
+            match message.delivery {
+                DeliveryRequirement::ReliableOrdered(Some(_)) => {
```

Comment:
```rust
  DeliveryRequirement::ReliableOrdered(stream)  | DeliveryRequirement::Default => {
                   if (stream.is_some()) {
                          warn!("Streams are not supported by TCP and will be ignored.");
                    }
                    write_message(message, &mut net);
                }
```

Maybe move this in a single match

Reply:
This can't be done in a single match because DeliveryRequirement::Default doesn't bind stream so it doesn't compile